docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1317,6 +1317,248 @@ As a Zephyr OS user I want to... ???,
 Architectural & interface REQ
 <<<
 
+[SECTION]
+TITLE: Counting Semaphores
+
+[REQUIREMENT]
+UID: ZEP-127
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Build-time registration for pre-kernel init phase notification
+STATEMENT: >>>
+Zephyr Counting Semaphores shall build-time register with Zephyr System Initialization for notification of running in the Zephyr Pre-Kernel initialization phase.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-128
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Count Semaphore pre-Kernel notification initializes statically defined semaphore objects
+STATEMENT: >>>
+When the Counting Semaphores service is notified of the Zephyr Pre-Kernel System Initialization phase,
+
+Zephyr Counting Semaphores shall initialize all statically defined semaphores.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-129
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore dynamic initialization
+STATEMENT: >>>
+Zephyr Counting Semaphores shall expose an interface to dynamically initialize a Counting Semaphore which requires the following information:
+
+- initial count of the Counting Semaphore; and
+- limit of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-130
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphores Static Object Init Failures Cause Build Errors
```

Comment:
this seems to be based on a BUILD_ASSERT that exists in K_SEM_DEFINE? Does this qualify for a dedicated requirement for each object we statically define or does this mean we will have a requirement for each BUILD_ASSERT in the code?

Maybe expand on the requirement above (ZEP-129) using EARS and say

".... WHERE the limit is larger or equal to the initial count and the limit is less the maximal limit specified by the system"

Error checking and paramter verificatio is a topic on its own, because Zephyr by design does not check every input and does does not fail gracefully (when not in userspace mode), see following comments...

Reply:
If the initialization parameters are invalid and the build completes without error it can result in unpredictable runtime behavior.  Checking the range of parameters during build is the same as verifying them when entering runtime initialization function.  Just good practice for safe coding.

There should be a global requirement that "All statically initialized objects have their parameters validated during build whenever possible."  Repeating the requirement where every it applies could create too much noise.

Reply:
@Timotheous is exactly right about the implicit higher-level requirement.

I've worked in an RM environment where the dictate was that requirements are complete and stand-alone. In that case there are no dependencies on document context. The advantage is downstream requirements have to link back to only one item per requirement and finding missing links becomes MUCH easier. The disadvantage is that statements propagate out. I wouldn't describe this repetition as noise. Every occurrence (and lack of occurrence) is important. Additionally, compound conditionals will exist. I don't think we can find an efficient organization for many of these compound conditions.  (I've already run across a triple-condition.)

[EARS syntax](https://alistairmavin.com/ears/) explicitly uses:
- WHERE for optional feature requirements;
- WHEN for state-driven requirements; and
- IF <trigger>, THEN for unwanted behavior requirements.

As far as I can tell this is fundamentally an unwanted behavior, secondarily a state, and certainly not an optional feature.

Yes, error checking and parameter verification is another topic that comes down from a higher layer.  Some regulations will impose policy. Some users will have very high priority needs in this area. Some behaviors will be left up to the Architecture Decisions. The most important point is unspecified behavior is implicitly insecure. IoT security regulations assessments require acceptable justifications for each non-compliant occurrence.